# Test PySpark & DataFrames

You will be asked some basic questions about some stock market data from the years 2012-2017. 

#### Use the stock.csv file to Answer and complete the  tasks below!

#### Start a Spark Session

In [ ]:
pip install py4j

     |████████████████████████████████| 198 kB 6.8 MB/s 


In [3]:
!pip install -q findspark
!pip install -q pyspark

     |████████████████████████████████| 281.4 MB 47 kB/s 
     |████████████████████████████████| 198 kB 62.0 MB/s 


In [1]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [2]:
# innstall java
!apt-get install openjdk-8-jdk-headless -qq > /dev/null
# install spark (change the version number if needed)
!wget -q https://archive.apache.org/dist/spark/spark-3.0.0/spark-3.0.0-bin-hadoop3.2.tgz
# unzip the spark file to the current folder
!tar xf spark-3.0.0-bin-hadoop3.2.tgz
# set your spark folder to your system path environment. 
import os
os.environ["JAVA_HOME"] = "/usr/lib/jvm/java-8-openjdk-amd64"
os.environ["SPARK_HOME"] = "/content/spark-3.0.0-bin-hadoop3.2"
# install findspark using pip
!pip install -q findspark

In [4]:
import findspark
findspark.init()

In [5]:
from pyspark.sql import SparkSession
spark = SparkSession.builder.appName('Basics').getOrCreate()

#### Load the Stock CSV File, have Spark infer the data types.

In [7]:
df = spark.read.csv('stock.csv',inferSchema=True,header=True)
df.show()

+----------+------------------+------------------+------------------+------------------+--------+------------------+
|      Date|              Open|              High|               Low|             Close|  Volume|         Adj Close|
+----------+------------------+------------------+------------------+------------------+--------+------------------+
|2012-01-03|         59.970001|         61.060001|         59.869999|         60.330002|12668800|52.619234999999996|
|2012-01-04|60.209998999999996|         60.349998|         59.470001|59.709998999999996| 9593300|         52.078475|
|2012-01-05|         59.349998|         59.619999|         58.369999|         59.419998|12768200|         51.825539|
|2012-01-06|         59.419998|         59.450001|         58.869999|              59.0| 8069400|          51.45922|
|2012-01-09|         59.029999|         59.549999|         58.919998|             59.18| 6679300|51.616215000000004|
|2012-01-10|             59.43|59.709998999999996|             5

#### List the 5th and 6th column names?

In [10]:
df.columns[4]


'Close'

In [ ]:
df.columns[5]

'Volume'

#### Print the Schema and change the inferred Volume type from integer to double

In [ ]:
df.printSchema()

root
 |-- Date: string (nullable = true)
 |-- Open: double (nullable = true)
 |-- High: double (nullable = true)
 |-- Low: double (nullable = true)
 |-- Close: double (nullable = true)
 |-- Volume: integer (nullable = true)
 |-- Adj Close: double (nullable = true)



In [ ]:
df.schema

StructType(List(StructField(Date,StringType,true),StructField(Open,DoubleType,true),StructField(High,DoubleType,true),StructField(Low,DoubleType,true),StructField(Close,DoubleType,true),StructField(Volume,IntegerType,true),StructField(Adj Close,DoubleType,true)))

In [23]:
from pyspark.sql.types import DoubleType

df_new = df.withColumn("Volume", df["Volume"].cast(DoubleType()))
df_new.printSchema()
df_new.select("Volume").show()

root
 |-- Date: string (nullable = true)
 |-- Open: double (nullable = true)
 |-- High: double (nullable = true)
 |-- Low: double (nullable = true)
 |-- Close: double (nullable = true)
 |-- Volume: double (nullable = true)
 |-- Adj Close: double (nullable = true)
 |-- HV Ratio: double (nullable = true)
 |-- LV Ratio: double (nullable = true)

+---------+
|   Volume|
+---------+
|1.26688E7|
|9593300.0|
|1.27682E7|
|8069400.0|
|6679300.0|
|6907300.0|
|6365600.0|
|7236400.0|
|7729300.0|
|8500000.0|
|5911400.0|
|9234600.0|
|1.03788E7|
|7134100.0|
|7362800.0|
|5915800.0|
|7436200.0|
|6287300.0|
|7636900.0|
|9761500.0|
+---------+
only showing top 20 rows



#### Print out the date and the open price of the first 5 rows as follows.
on 2012-01-03 00:00:00 , the open price was 59.97 $


on 2012-01-04 00:00:00 , the open price was 60.21 $


on 2012-01-05 00:00:00 , the open price was 59.35 $


on 2012-01-06 00:00:00 , the open price was 59.42 $


on 2012-01-09 00:00:00 , the open price was 59.03 $

In [22]:
print("on the ", df.first()['Date'], ", the open price was ", df.first()['Open'], "$")

 # df.collect row , columns


on the  2012-01-03 , the open price was  59.970001 $


#### Use describe() to generate a summary DataFrame.

In [24]:
df.describe()

DataFrame[summary: string, Date: string, Open: string, High: string, Low: string, Close: string, Volume: string, Adj Close: string, HV Ratio: string, LV Ratio: string]

In [30]:
df_describe = df.describe().show()
df_describe


+-------+----------+------------------+-----------------+-----------------+-----------------+-----------------+-----------------+--------------------+--------------------+
|summary|      Date|              Open|             High|              Low|            Close|           Volume|        Adj Close|            HV Ratio|            LV Ratio|
+-------+----------+------------------+-----------------+-----------------+-----------------+-----------------+-----------------+--------------------+--------------------+
|  count|      1258|              1258|             1258|             1258|             1258|             1258|             1258|                1258|                1258|
|   mean|      null| 72.35785375357709|72.83938807631165| 71.9186009594594|72.38844998012726|8222093.481717011|67.23883848728146|1.058087386431558E-5|1.046022660417144...|
| stddev|      null|  6.76809024470826|6.768186808159218|6.744075756255496|6.756859163732991|  4519780.8431556|6.722609449996857| 4.34985596

AttributeError: ignored

#### There are too many decimal places for mean and stddev in the describe() dataframe. Format the numbers to just show up to two decimal places. Pay careful attention to the datatypes that .describe() returns. [Check this link for a hint](https://spark.apache.org/docs/latest/api/python/reference/api/pyspark.sql.Column.cast.html?highlight=cast)

In [38]:
from pyspark.sql.types import IntegerType


#### Create a new dataframe with 2 columns : 
##### HV Ratio that is the ratio of the High Price versus volume of stock traded for a day,  
##### and LV Ratio that is the ratio of the Low Price versus volume of stock traded for a day.

In [19]:
from pyspark.sql.functions import col, lit
df = df.withColumn('HV Ratio',col("High")/col("Volume"))
df = df.withColumn('LV Ratio',col("Low")/col("Volume"))
df.select('HV Ratio', 'LV Ratio').show()

+--------------------+--------------------+
|            HV Ratio|            LV Ratio|
+--------------------+--------------------+
|4.819714653321546E-6|4.725782947082596...|
|6.290848613094555E-6|6.199118238770809E-6|
|4.669412994783916E-6|4.571513525790636E-6|
|7.367338463826307E-6|7.295461744367611E-6|
|8.915604778943901E-6|8.821283368017606E-6|
|8.644477436914568E-6|8.538792292212586E-6|
|9.351828421515645E-6|9.274852488375017E-6|
| 8.29141562102703E-6| 8.20850174119728E-6|
|7.712212102001476E-6|7.634585020635763E-6|
|7.071764823529412E-6|7.002352941176471E-6|
|1.015495466386981E-5|1.009067259870758...|
|6.576354146362592...|6.470231520585624E-6|
| 5.90145296180676E-6|5.845569622692411E-6|
|8.547679455011844E-6|8.481798404844339E-6|
|8.420709512685392E-6|8.250937143478025E-6|
|1.041448341728929...|1.031813127556712...|
|8.316075414862431E-6| 8.17218471800113E-6|
|9.721183814992126E-6|9.628934677842636E-6|
|8.029436027707578E-6|7.902420877581218E-6|
|6.307432259386365E-6|6.20601362

#### What day had the Highest HV Ratio? and what day we had the lowest?

In [35]:
from pyspark.sql.functions import mean, min, max

max = df.select([max("HV Ratio")]).show()


+--------------------+
|       max(HV Ratio)|
+--------------------+
|3.734784381116044E-5|
+--------------------+



In [36]:
min = df.select([min("HV Ratio")]).show()

+--------------------+
|       min(HV Ratio)|
+--------------------+
|8.399455240605156E-7|
+--------------------+



#### What is the mean of the LV Ratio?

In [37]:
mean = df.select([mean("LV Ratio")]).show()

+--------------------+
|       avg(LV Ratio)|
+--------------------+
|1.046022660417144...|
+--------------------+



#### How many days volume was greater than 9000000 and the Close was lower than 70$

In [51]:
df_b = df.filter(df['Volume'] > 9000000 ) 
df_b = df_b.filter(df_b['Close'] < 70 )
df_b.count()

208

#### What percentage of the time was the difference between High and Low greater than 1 dollars ?

30.206677265500797

#### What is the Pearson correlation between Close and Volume?
#### [Hint](https://spark.apache.org/docs/latest/api/python/reference/pyspark.pandas/api/pyspark.pandas.DataFrame.corr.html?highlight=corr)

+--------------------+
| corr(Close, Volume)|
+--------------------+
|-0.36864129499157394|
+--------------------+



#### What is the max and min HV Ratio per year?

+----+--------------------+
|Year|       max(HV Ratio)|
+----+--------------------+
|2015|2.460931126147897...|
|2013|3.734784381116044E-5|
|2014|3.494261176659442...|
|2012|2.368837716882401...|
|2016|1.712227308529861...|
+----+--------------------+

+----+--------------------+
|Year|       min(HV Ratio)|
+----+--------------------+
|2015|8.399455240605156E-7|
|2013|2.725464010247744...|
|2014|3.641002174256106E-6|
|2012|1.584958731612085...|
|2016|1.976525784922755E-6|
+----+--------------------+



#### What is the highest Close for each Calendar Month?
#### In other words, across all the years, what is the highest Close price for Jan,Feb, Mar, etc... Your result will have a value for each of these months. 

+-----+-----------------+
|Month|       max(Close)|
+-----+-----------------+
|    1|        90.470001|
|    2|87.33000200000001|
|    3|        83.959999|
|    4|        81.029999|
|    5|79.91999799999999|
|    6|            77.32|
|    7|        78.550003|
|    8|        78.769997|
|    9|        77.510002|
|   10|        78.290001|
|   11|        87.540001|
|   12|        86.910004|
+-----+-----------------+

